### DISCUSSION TITLE
Collectable in-map chaos emeralds

### DISCUSSION THREAD

**Discussion by saneway** [Techokami/SonicWorldsNext#118] [Q&A] [answered]
How would one go about making these? I've tried getting them to work myself but I'm a bit unfamiliar with enums and bit flags, even after reading up on them in the Godot documentation.

Edit: For clarity's sake, I have made a collectable chaos emerald object, I am just unsure on how to update the emerald count itself upon collecting it

**Comment by DimensionWarped** [comment, marked as answer]
Global.emeralds is the value you are looking for and it's a bitmask IE

00000000 - 0 - the player has no emeralds
00000101 - 5 - the player has emeralds 1 and 3 (missed 2)
01100101 - 101 - the player has emerals 1, 3, 6, and 7
01111111 - 127  - the player has all seven emeralds

If you want to map that out to decimal values, emerald 1 is 1, emerald 2 is 2, emerald 3 is 4, emerald 4 is 8, emerald 5 is 16, emerald 6 is 32, emerald 7 is 64 - if you aren't familiar with bitwise math, this might seem tricky, but if you simply add 16 to whatever the value is (assuming you haven't done it before, in which case this will break), you'd be putting that emerald (emerald 5) in the player's "inventory".

I'd strongly recommend you either learn a little bitwise operators before trying to use this. Alternatively you could just reimplement this yourself in a way that you feel more comfortable with (like an array of Booleans if that makes more sense to you)

https://www.youtube.com/watch?v=igIjGxF2J-w&t=3s

You can look SpecialStageResult.gd for examples of it being manipulated
